contributors/design-proposals/scheduling/gang-scheduling.md

```diff
@@ -0,0 +1,174 @@
+
+**Authors**: @k82cn , **Shepherds**: @vishh, @bsalamat
+
+**SIG/WG**: sig-scheduling, ML-WG
+
+## Motivation
+
+After the discussion at [Gang-scheduling](https://docs.google.com/document/d/1AUwcvTtULNvow5M9e428FnlvINO1uQ7ojRoTGuTp4DA/edit#heading=h.ckn8nv2jj0xv) proposal, we decide to make Gang-scheduling API out of core by CRD, and implement it in [kube-arbitrator](https://github.com/kubernetes-incubator/kube-arbitrator). kube-arbitrator focus on "batch" workload in kubernetes, and will share the same [scheduling frameworks](https://github.com/kubernetes/community/pull/2281) when it's ready. This document is used to provide definition of CRD for gang-scheduling, and suggestion of customized job controllers.
+
+## Function Detail
+
+### CRD Definition
+
+Although workload lifecycle requirements maybe different, e.g. MPI vs. Spark, the requirements to scheduler are similar. To meet both scheduling and lifecycle requirements, two **Kinds** are going to be defined. One new **Kind**, named `SchedulingSpec`, is defined as follow for scheduler:
+
+```go
+// SchedulingSpec defines the scheduling requirement of a 'Job'
+type SchedulingSpec struct {
+    metav1.TypeMeta
+    metav1.ObjectMeta
+
+    Spec SchedulingSpecTemplate
+}
+
+// SchedulingSpecTemplate represents the template of SchedulingSpec.
+type SchedulingSpecTemplate struct {
+    // If specified, indicates the Job's priority. "system-node-critical" and
+    // "system-cluster-critical" are two special keywords which indicate the
+    // highest priorities with the former being the highest priority. Any other
+    // name must be defined by creating a PriorityClass object with that name.
+    // If not specified, the Job's priority will be default or zero if there is no
+    // default.
+    // +optional
+    PriorityClassName string
+
+    // The priority value. Various system components use this field to find the
+    // priority of the Job. When Priority Admission Controller is enabled, it
+    // prevents users from setting this field. The admission controller populates
+    // this field from PriorityClassName.
+    // The higher the value, the higher the priority.
+    // +optional
+    Priority *int32
+
+    // NodeSelector is a selector which must be true for the pod of 'Job' to
+    // fit on a node. Selector which must match a node's labels for the pod of
+    // 'Job' to be scheduled on that node.
+    // +optional
+    NodeSelector map[string]string
+
+    // MinAvailable defines the minimal available tasks to run the Job;
+    // if there's not enough resources to start all tasks, the scheduler
+    // will not start anyone.
+    MinAvailable int
+}
+```
+
+The `SchedulingSpec` defines necessary parameters of Job for scheduler, e.g. priority of Job, minimal available tasks of Job. `SchedulingSpec` does not include lifecycle parameters which are managed by following new **Kind**, named `QueueJob`:
+
+```go
+// QueueJob is a Kind for batch workload.
+type QueueJob struct {
```

Comment:
If you were to consider the approach @bsalamat suggested, we would not need a queue job I think.

Reply:
Yes; if so, I'd like to enhance 'Job' instead of introducing a new Kind.

Reply:
For example, multiple pod template (or wrapper of multiple job).